### PR TITLE
[fuchsia] Adds a singleton to access inspect root node.

### PR DESCRIPTION
**Pull request by lucasradaelli**
Bug:  fxbug.dev/75281

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue.*
fxbug.dev/75281

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: shell/platform/fuchsia/flutter/main.cc [hunk 1]
```diff
@@ -12,20 +12,20 @@
...
   auto context = sys::ComponentContext::Create();
-  sys::ComponentInspector inspector(context.get());
-
-  inspect::Node& root = inspector.inspector()->GetRoot();
+  dart_utils::RootNodeSingleton::Initialize(context.get());
 
   // We inject the 'vm' node into the dart vm so that it can add any inspect
   // data that it needs to the inspect tree.
-  dart::SetDartVmNode(std::make_unique<inspect::Node>(root.CreateChild("vm")));
+  dart::SetDartVmNode(std::make_unique<inspect::Node>(
```
Comment (chaselatta) [inline shell/platform/fuchsia/flutter/main.cc:27]: can you also do the same in the dart runner's main?

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// found in the LICENSE file.
+
+#include "root_node_singleton.h"
+
+namespace dart_utils {
+
+std::unique_ptr<sys::ComponentInspector> RootNodeSingleton::inspector_;
+std::mutex RootNodeSingleton::mutex_;
+
+void RootNodeSingleton::Initialize(sys::ComponentContext* context) {
```
Comment (chaselatta) [inline shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc:12]: How do you feel about just calling this something like RootInspectNodeSingleton or RootInspectNode? The name node is a bit generic so we want to make it clear that this is for inspect.

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.h [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include <memory>
+#include <mutex>
+
+namespace dart_utils {
+
+// This singleton object offers a way to create a new inspect node under the
+// root node in a thread safe way.
+//
+// Usage notes:
+// RootNodeSingleton ::Initialize() must  be invoked once in the program's main
```
Comment (chaselatta) [inline shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.h:19]: nit: did you mean to have an extra space between RootNodeSingleton and ::

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc (hunk 2 above)
Comment (arbreng) [reply shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc:12]: +10 for removing Singleton from the name, I like RootInspectNode

File: shell/platform/fuchsia/flutter/main.cc (hunk 1 above)
Comment (lucasradaelli) [reply shell/platform/fuchsia/flutter/main.cc:27]: done.

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc (hunk 2 above)
Comment (lucasradaelli) [reply shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc]: Done.

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.h (hunk 3 above)
Comment (lucasradaelli) [reply shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.h]: done.

File: shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc (hunk 2 above)
Comment (lucasradaelli) [reply shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc]: will remove the singleton -- just saw the comment after uploading.

Comment (lucasradaelli) [reply shell/platform/fuchsia/runtime/dart/utils/root_node_singleton.cc]: done.

**Review by chaselatta — APPROVED** [review]

**Comment by lucasradaelli** [comment]
I don't have write access to submit this.

- Should I apply somewhere to obtain it?
- - Or should someone else submit this for me?

**Review by arbreng — APPROVED** [review]

File: shell/platform/fuchsia/runtime/dart/utils/root_inspect_node_singleton.cc [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+void RootInspectNodeSingleton::Initialize(sys::ComponentContext* context) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  if (!inspector_) {
+    inspector_ = std::make_unique<sys::ComponentInspector>(context);
+  }
+}
+
+inspect::Node RootInspectNodeSingleton::CreateRootChild(
+    const std::string& name) {
+  std::lock_guard<std::mutex> lock(mutex_);
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/runtime/dart/utils/root_inspect_node_singleton.cc]: See:  https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=75534

**Pull request merged**